Modify OrderBy Fluent To Build On Collect

a046 opened the pull request:
Open for suggestions on how to avoid having to make CollectPatternBuilder.cs, it just seemed like a nicer way to get the AggregateBuilder+SourceBuilder than exposing them on the PatternBuilder/AggregateBuilder classes and/or adding extra parameters in QueryExpressions.

For #171

a046 committed:
Modify OrderBy Fluent To Build On Collect

snikolayev commented on src/NRules/NRules.Fluent/Dsl/QueryExpression.cs:
```diff
@@ -20,4 +22,20 @@ public QueryExpression(IQueryBuilder builder)
+    /// Expression builder for collect queries.
+    /// </summary>
+    /// <typeparam name="TSource">Type of query source.</typeparam>
+    public class CollectQueryExpression<TSource> : QueryExpression<IEnumerable<TSource>>, ICollectQuery<TSource>
```
This is not really needed, as it serves no purpose - `QueryExpression<T>` can implement both, `IQuery<T>` and `ICollectQuery<T>`

a046 replied:
I'll do that, but then it becomes ICollectQuery<IEnumerable<TSource>> which I don't think is as nice.

It's also not as type safe as someone could cast something else to ICollectQuery, but given the way these classes are used I guess it's not a major concern.

snikolayev replied:
Ok, I may have overlooked the relationship between IQuery and ICollectQuery. I think we want type safety. Either way, its a quick thing to change one way or the other, just need to get a feel for it to see which one makes more sense. Perhaps you are right.

snikolayev replied:
Thinking about it more, I think it's fine. The purpose of this addl interface is to drive what extension methods are available, and the type information just flows through as normal.

snikolayev commented on src/NRules/NRules.Fluent/Dsl/QueryExtensions.cs:
```diff
@@ -142,10 +142,10 @@ public static IQuery<TSource> Where<TSource>(this IQuery<TSource> source, params
         {
             source.Builder.Collect<TSource>();
-            return new QueryExpression<IEnumerable<TSource>>(source.Builder);
+            return new CollectQueryExpression<TSource>(source.Builder);
```
This would just return QueryExpression, as long as that implements ICollectQuery, see my previous comment.

snikolayev commented on src/NRules/NRules.RuleModel/Builders/CollectPatternBuilder.cs:
```diff
@@ -0,0 +1,38 @@
+    /// <summary>
+    /// Builder to compose a Collect rule pattern.
+    /// </summary>
+    public class CollectPatternBuilder : PatternBuilder
```
Instead of a `CollectPatternBuilder`, I would handle this fully inside `QueryExpression`.
Instead of returning a `PatternBuilder` from the `_buildAction`, it should return a new type, something like:
```c#
private class AggregateBuildResult
{
    public PatternBuilder Pattern { get; }
    public AggregateBuilder Aggregate { get; }
    public PatternBuilder Source { get; }
}
```
Then you can manipulate the `AggregateBuilder` inside the `OrderBy` to call `Sort` on it.

snikolayev reviewed (changes requested):
Thanks! I have an idea about how to deal with this w/o the CollectPatternBuilder.
Another thing is that perhaps you no longer need a Sort type of aggregator in AggregateElement, as we discussed before, and just treat it as a Collect aggregator, but with additional expressions.

a046 committed:
Address review comments

a046 committed:
Merge pull request #1 from a046/fix-fluent

Address review comments

a046 commented:
Hmm, my visual studio xunit runner is broken locally -- wasn't able to run them locally before submitting to AppVeyor. I'll look at that failure tomorrow, sorry!

snikolayev commented:
It fails because in ReteBuilder you are now always building a CollectionAggregatorFactory even if sorting key selectors are present. So, most unit tests for sorted aggregates fail.

a046 commented:
Thanks, got it working and noticed that.

I've changed CollectAggregatorFactory to also construct the SortedAggregators and deleted the SortedAggregatorFactory which fixed all the tests. Is that how you were thinking it would work?

snikolayev commented:
That's fine too. I was thinking that you would choose between aggregator factories in ReteBuilder based on what expressions are available, but perhaps what you did is better, as it encapsulates that logic inside the collection aggregator factory. So, sounds good.

a046 committed:
Fix sorted aggregator factory
